File: MatlabOctaveCodeSnippets/snip_systems_phase_estimation.m
[h,Fs,b]=readwav('impulseResponses.wav');
nstart=11026; %when second \delta[n] occurs in impulses.wav
nend=22050; %segment ends before the third \delta[n] in impulses.wav
h=double(h(nstart:nend)); %segment and cast h to double
N=length(h)-1; %it's convenient to make N an even number to use N/2
H=fft(h,N); %calculate FFT
p = unwrap(angle(H(1:N/2+1))); %unwrap the phase for positive freqs.
f=Fs/N*(0:N/2)/1000; %create abscissa in kHz. Fs/N is the bin width
handler=plot(f,p); xlabel('f (kHz)'),ylabel('\angle{H(f)} (rad)')
k1=1500; k2=4000; makedatatip(handler,[k1,k2]); %choose any 2 points
%Calculate derivative as the slope. Convert from kHz to rad/s first:
groupDelayInSeconds = -atan2(p(k2)-p(k1),2*pi*1000*(f(k2)-f(k1)))

File: PythonCodeSnippets/snip_systems_phase_estimation.py
from scipy.io import wavfile
import numpy as np
import matplotlib.pyplot as plt

# Read the wav file
Fs, h = wavfile.read("impulseResponses.wav")

# when second \delta[n] occurs in impulses.wav
nstart = 11026

# segment ends before the third \delta[n] in impulses.wav
nend = 22050

# segment and cast h to double
h = h[nstart:nend].astype(float)

# it's convenient to make N an even number to use N/2
N = len(h) - 1

# calculate FFT
H = np.fft.fft(h, N)

# unwrap the phase for positive freqs.
p = np.unwrap(np.angle(H[: N // 2 + 1]))

# create abscissa in kHz. Fs/N is the bin width
f = Fs / N * np.arange(0, N / 2 + 1) / 1000

# plot
plt.plot(f, p)
plt.xlabel("f (kHz)")
plt.ylabel("angle{H(f)} (rad)")

# choose any 2 points
k1 = 1500
k2 = 4000

# Calculate derivative as the slope. Convert from kHz to rad/s first:
groupDelayInSeconds = -np.arctan2(p[k2] - p[k1], 2 * np.pi * 1000 * (f[k2] - f[k1]))
